docs/NNNN-direct-clean-path-layout.md

```diff
@@ -0,0 +1,451 @@
+* **`maxFilenameLen`/`maxPathnameLen`:**
+  Several filesystems (e.g. Ext2/3/4) have byte restrictions on filename length.
+  When using UTF16 (i.e. NTFS) or UTF32 characters in the filesystem, the byte length is double or quad of the character length.
+  For filesystems which are using UTF8 character sets, the length of a name in bytes 
+  can be calculated only by building the UTF8 string and count the byte length afterwards.
+
+  **Hint:** UTF8 has always less or equal bytes than UTF32 which means, that
+  assuming UTF32 instead of UTF8 for length calculation is safe, but would give you
+  only 63 characters on 255 byte restrictions.
+  
+## Procedure
+
+The following is an outline to the steps for mapping an identifier/filepath.
+    
+[UTF Replacement Character List](#utf-replacement-character-list)
+is a list of UTF characters mentioned below.
+
+### When `utfEncode` is `true`
+
+1. Replace all non-UTF8 characters with `replacementString`
+2. Split the string at path separator `/`
+3. For each part do the following
+   1. Replace `=` with `=u003D` if it is followed by `u` and four hex digits 
+   2. Replace any character from this list with its utf code in the form `=uXXXX` 
+       where `XXXX` is the code: 
+       `U+0000-U+001F` `U+007F` `U+0020` `U+0085` `U+00A0` `U+1680` `U+2000-U+200F` 
+       `U+2028` `U+2029` `U+202F` `U+205F` `U+3000` `\n` `\t` `*` `?` `:` `[` `]` `"` 
+       `<` `>` `|` `(` `)` `{` `}` `&` `'` `!` `;` `#` `@`  
+   3. If part only contains periods, replace first period (`.`), with UTF Code (`U+002E`)
+   4. Remove part completely, if its length is 0
+   5. When length of part is larger than `maxFilenameLen` use fallback function and return
+4. Join the parts with path separator `/`
+5. When length of result is larger than `maxPathnameLen` use fallback function and return
+
+### When `utfEncode` is `false`
+
+1. Replace all non-UTF8 characters with `replacementString`
+2. Split the string at path separator `/`
+3. For each part do the following
+    1. Replace any whitespace character from this list with `whitespaceReplacementString`: 
+       `U+0009` `U+000A-U+000D` `U+0020` `U+0085` `U+00A0` `U+1680` `U+2000-U+200F`
+       `U+2028` `U+2029` `U+202F` `U+205F` `U+3000`
+    2. Replace any character from this list with `replacementString`: 
+       `U+0000-U+001F` `U+007f` `*` `?` `:` `[` `]` `"`
+       `<` `>` `|` `(` `)` `{` `}` `&` `'` `!` `;` `#` `@`
+    3. Remove leading spaces, `-` and `~` / remove trailing spaces
+    4. If part only contains periods, replace first period (`.`) with `replacementString`
+    5. Remove part completely, if its length is 0
+    6. When length of part is larger than `maxFilenameLen` use fallback function and return
+4. Join the parts with path separator `/`
+5. When length of result is larger than `maxPathnameLen` use fallback function and return
+
+### Fallback function
+
+1. Create digest with `fallbackDigestAlgorithm` from initial parameter as lower case hex string
```

Comment:
Can you elaborate on what you are referring to with: "initial parameter"?

Reply:
replaced parameter with identifier/filepath